### PR TITLE
feat(protocol): adopt optimism new trie codebase

### PR DESCRIPTION
**Pull request by dantaik**
From Brecht:

> We may want to update the Merkle trie/RLP decoding to the latest versions used in optimism: https://github.com/ethereum-optimism/optimism/tree/develop/packages/contracts-bedrock/src/libraries. I haven't diffed them, but seems likely there are also some optimizations in there. In any case, we should definitely check if the new versions contain fixes or improvements.

### CODE REVIEW

File: packages/protocol/contracts/L1/hooks/AssignmentHook.sol [hunk 1]
```diff
@@ -182,8 +182,11 @@ contract AssignmentHook is EssentialContract, IHook {
         pure
         returns (uint256)
     {
-        for (uint256 i; i < tierFees.length; ++i) {
+        for (uint256 i; i < tierFees.length;) {
             if (tierFees[i].tier == tierId) return tierFees[i].fee;
+            unchecked {
+                ++i;
+            }
```
Comment (adaki2004) [inline packages/protocol/contracts/L1/hooks/AssignmentHook.sol:189]: No need to have this optimizaton, bc. fro 0.8.22:

kép

Unchecked loop increemnt optimizer:
kép
https://docs.soliditylang.org/en/latest/internals/optimizer.html#unchecked-loop-optimizer

File: packages/protocol/contracts/L1/libs/LibProposing.sol [hunk 2]
```diff
@@ -260,6 +260,9 @@ library LibProposing {
                 );
 
                 prevHook = params.hookCalls[i].hook;
+                unchecked {
+                    ++i;
+                }
```
Comment (adaki2004) [inline packages/protocol/contracts/L1/libs/LibProposing.sol:265]: Same as here:
https://github.com/taikoxyz/taiko-mono/pull/15608/files#r1470670145

File: packages/protocol/contracts/L1/provers/Guardians.sol [hunk 3]
```diff
@@ -57,8 +57,11 @@ abstract contract Guardians is EssentialContract {
 
         // Delete current guardians data
         uint256 guardiansLength = guardians.length;
-        for (uint256 i; i < guardiansLength; ++i) {
+        for (uint256 i; i < guardiansLength;) {
             delete guardianIds[guardians[i]];
+            unchecked {
+                ++i;
+            }
```
Comment (adaki2004) [inline packages/protocol/contracts/L1/provers/Guardians.sol:64]: Same as here:
https://github.com/taikoxyz/taiko-mono/pull/15608/files#r1470670145

File: packages/protocol/contracts/signal/SignalService.sol [hunk 4]
```diff
@@ -168,4 +171,18 @@ contract SignalService is AuthorizableContract, ISignalService {
...
+    /// TrieNodes.
+    function _transcode(bytes memory proof) internal pure returns (bytes[] memory proofs) {
+        RLPReader.RLPItem[] memory nodes = RLPReader.readList(proof);
+        proofs = new bytes[](nodes.length);
+
+        for (uint256 i; i < nodes.length;) {
+            proofs[i] = RLPReader.readBytes(nodes[i]);
+            unchecked {
+                ++i;
+            }
```
Comment (adaki2004) [inline packages/protocol/contracts/signal/SignalService.sol:185]: Same as here: https://github.com/taikoxyz/taiko-mono/pull/15608/files#r1470670145

File: packages/protocol/contracts/L1/hooks/AssignmentHook.sol (hunk 1 above)
Comment (dantaik) [reply packages/protocol/contracts/L1/hooks/AssignmentHook.sol:189]: Good to know.

**Review by Brechtpd — APPROVED** [review]

File: packages/protocol/contracts/signal/SignalService.sol [hunk 5]
```diff
@@ -170,6 +173,17 @@ contract SignalService is AuthorizableContract, ISignalService {
...
+    /// @notice Translate a RLP-encoded list of RLP-encoded TrieNodes into a list of LP-encoded
+    /// TrieNodes.
+    function _transcode(bytes memory proof) internal pure returns (bytes[] memory proofs) {
+        RLPReader.RLPItem[] memory nodes = RLPReader.readList(proof);
+        proofs = new bytes[](nodes.length);
+
+        for (uint256 i; i < nodes.length; ++i) {
+            proofs[i] = RLPReader.readBytes(nodes[i]);
+        }
+    }
```
Comment (adaki2004) [inline packages/protocol/contracts/signal/SignalService.sol:185]: Obviously prefer not having this on-chain (and i know Daniel already opened a PR ).

We can tho manage to upgrade in a ways that: X messageId is old version implementation verification - after X this new ?
So basically having them living both side by side for some time.. Tehn upgrade the contracts again and exclude the old.

This way we can be compatible by not causing too much damage.

Comment (dantaik) [reply packages/protocol/contracts/signal/SignalService.sol:185]: We can do it in two PRs, after merging this one,  changing this PR to support the tokenID based handling  --https://github.com/taikoxyz/taiko-mono/pull/15609. The old code can still be deleted.

Comment (dantaik) [reply packages/protocol/contracts/signal/SignalService.sol:185]: Wait, the `messageId` is only visble on the Bridge level, not on the SignalService level, so we have to figure out another way ...

**Review by davidtaikocha — APPROVED** [review]

File: packages/protocol/contracts/signal/SignalService.sol (hunk 5 above)
Comment (dantaik) [reply packages/protocol/contracts/signal/SignalService.sol:185]: I'll simply ask Korbi to use eth_getstorageat to get the SignalService's implementation address, then if address == old_impl, then use the relayer's returned proof as-is, if the addrss == new_impl, the bridge dapp should RLP decode then proof, then ABI-encode the decoded list, then use the generated data.

**Review by adaki2004 — APPROVED** [review]

**Pull request merged**